cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -83,20 +82,30 @@ def process_data(old_data, new_data)
       new_data.reverse_each do |stackname, data|
         reader = data[:loader].reader
         set_env_variables(stackname, current_grid) # set envs for execution time
-        data[:stack_data] = reader.execute(
-          values: data[:variables],
-          defaults: old_data[stackname].nil? ? nil : old_data[stackname][:stack_data]['variables'],
+        values = data[:variables]
+        if old_stack = old_data[stackname]
+          values = values.merge(old_stack[:stack_data]['variables'])
+        end
+        parsed_stack = reader.execute(
+          values: values,
           parent_name: data[:parent_name],
           name: data[:name]
         )
+        parsed_stack["parent"] = { "name" => parsed_stack.delete("parent_name") }
```

Comment:
This is here so that old and new data are somewhat similar.

Reply:
This changes what the server `PUT /v1/stacks/...` API sees from `{"parent_name": ...}` to `{"parent": {"name": ...}}`, including also `{"parent":{"name":null}}` for the top-level stack.

It works because the server `Stacks::Update` isn't expecting any `parent` or `parent_name` field, so both of them get ignored. Still feels a little wrong somehow.

Doesn't break the actual `POST /v1/grids/.../stacks` => `Stacks::Create` API, because that's from `kontena stack install --parent-name=...`

Reply:
It feels a little wrong that post accepts `parent_name` but response contains `parent` hash. I can remove whole parent thing from `PUT` because it's not used there at all.